src/appleseed/renderer/kernel/lighting/backwardlightsampler.cpp

```diff
@@ -613,29 +490,44 @@ float LightSampler::evaluate_pdf(const ShadingPoint& shading_point) const
         shading_point.get_primitive_index());
 
     const EmittingTriangle* triangle = m_emitting_triangle_hash_table.get(triangle_key);
-    return triangle->m_triangle_prob * triangle->m_rcp_area;
+
+    // Traverse the tree and update triangle_prob
+    float triangle_probability;
+    m_light_tree.light_probability(
+        shading_point.get_point(),
+        triangle->m_light_tree_node_index,
+        triangle_probability);
+
+    return triangle_probability * triangle->m_rcp_area;
 }
 
-void LightSampler::sample_light_tree_light(
+void BackwardLightSampler::sample_light_tree_light(
     const ShadingRay::Time&             time,
+    const Vector2f&                     s,
+    const int                           light_type,
     const size_t                        light_index,
     const float                         light_prob,
     LightSample&                        light_sample) const
 {
-    // Fetch the light.
-    const NonPhysicalLightInfo& light_info = m_light_tree_lights[light_index];
-    light_sample.m_light = light_info.m_light;
-
-    // Evaluate and store the transform of the light.
-    light_sample.m_light_transform =
-          light_info.m_light->get_transform()
-        * light_info.m_transform_sequence.evaluate(time.m_absolute);
-
-    // Store the probability density of this light.
-    light_sample.m_probability = light_prob;
+    if (light_type == LightSource::NonPhysicalLightType)
+    {
+        // Fetch the light.
+        const NonPhysicalLightInfo& light_info = m_light_tree_lights[light_index];
+        light_sample.m_light = light_info.m_light;
+    
+        // Evaluate and store the transform of the light.
+        light_sample.m_light_transform =
+              light_info.m_light->get_transform()
+            * light_info.m_transform_sequence.evaluate(time.m_absolute);
+    
+        // Store the probability density of this light.
+        light_sample.m_probability = light_prob;
+    }
+    else if (light_type == LightSource::EmittingTriangleType)
```

Comment:
Can it be anything else?

If not, I suggest removing the `if` and replacing it by an assertion:

    else
    {
        assert(light_type == LightSource::EmittingTriangleType);
        ...
    }